pkg/controller/directimagemigration/validate.go

```diff
@@ -0,0 +1,273 @@
+	ref := imageMigration.Spec.DestMigClusterRef
+
+	if !migref.RefSet(ref) {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     InvalidDestinationClusterRef,
+			Status:   True,
+			Reason:   NotSet,
+			Category: Critical,
+			Message:  InvalidDestinationClusterReferenceMessage,
+		})
+		return nil
+	}
+
+	// Check if clusters are unique
+	if reflect.DeepEqual(ref, imageMigration.Spec.SrcMigClusterRef) {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     InvalidDestinationCluster,
+			Status:   True,
+			Reason:   NotDistinct,
+			Category: Critical,
+			Message:  InvalidDestinationClusterMessage,
+		})
+		return nil
+	}
+
+	cluster, err := migapi.GetCluster(r, ref)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	// Not found
+	if cluster == nil {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     InvalidDestinationClusterRef,
+			Status:   True,
+			Reason:   NotFound,
+			Category: Critical,
+			Message:  InvalidDestinationClusterReferenceMessage,
+		})
+		return nil
+	}
+
+	// Not ready
+	if !cluster.Status.IsReady() {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     DestinationClusterNotReady,
+			Status:   True,
+			Category: Critical,
+			Message:  DestinationClusterNotReadyMessage,
+		})
+	}
+	// Exposed registry path
+	registryPath, err := cluster.GetRegistryPath(r)
+	if err != nil || registryPath == "" {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     MissingDestinationClusterRegistryPath,
+			Status:   True,
+			Category: Critical,
+			Message:  MissingDestinationClusterRegistryPathMessage,
+		})
+	}
+	return nil
+}
+
+// Validate required namespaces on the source cluster.
+// Returns error and the total error conditions set.
+func (r ReconcileDirectImageMigration) validateNamespaces(imageMigration *migapi.DirectImageMigration) error {
+	count := len(imageMigration.Spec.Namespaces)
+	if count == 0 {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     NsListEmpty,
+			Status:   True,
+			Category: Critical,
+			Message:  NsListEmptyMessage,
+		})
+		return nil
+	}
+	cluster, err := imageMigration.GetSourceCluster(r)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+	if cluster == nil || !cluster.Status.IsReady() {
+		return nil
+	}
+	client, err := cluster.GetClient(r)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+	ns := kapi.Namespace{}
+	notFound := make([]string, 0)
+	for _, nsName := range imageMigration.GetSourceNamespaces() {
+		err := client.Get(context.TODO(), types.NamespacedName{Name: nsName}, &ns)
+		if err == nil {
+			continue
+		}
+		if k8serror.IsNotFound(err) {
+			notFound = append(notFound, nsName)
+		} else {
+			return liberr.Wrap(err)
+		}
+	}
+	if len(notFound) > 0 {
+		imageMigration.Status.SetCondition(migapi.Condition{
+			Type:     NsNotFoundOnSourceCluster,
+			Status:   True,
+			Reason:   NotFound,
+			Category: Critical,
```

Comment:
If there are 10 items in the list of namespaces and 1 namespace is not present, I would expect the controller to migrate imagestreams for other 9 namespaces, two questions:

1. Is that a reasonable expectation?
2. If the above answer is yes, looks like the code is not doing that currently because Critical condition will block running the itinerary. Should this be modified?

Reply:
I was following the current behavior of the migplan validation -- if we say "migrate a namespace" and that namespace doesn't exist, then we fail validation before we start, assuming the user made a mistake, possibly a namespace name typo. If we remove "namespace existence" validation and choose to filter instead, that would work, but it changes the behavior in a way that needs to be documented (although, really, none of this is documented now). If I migrate a single namespace and that namespace doesn't exist, does this mean that it's a successful (no-op) migration? Failing on validation before running feels like the "fail early" approach, letting the user know now that a namespace they think they're migrating doesn't exist, rather than silently omitting it, and they find out at the end that something expected to be there isn't.

Reply:
I do not mean to suggest silently omitting, we definitely want to show the namespaces that do not exist, but would that block migration of other namespaces if my question?

In general, I am not keeping the practices adopted in migplan controller while reviewing this controller PR. In some cases, the migplan controller is intended to serve the UI and our practices evolved with that need. This new controller is more internal and would set my expectations a little differently

Reply:
Right. I mention the migplan controller as the starting point not because we can't change it, but because it's the starting point. If there's a good reason to do differently, we can. It seems to me that failing early and letting the user either remove or fix the nonexistent namespace is the more intuitive approach, but we could handle it differently. If I'm understanding it, the basic part of your suggestion is to remove this from the pre-migration validation and to treat it as a runtime warning or error. Instead of blocking migration completely, we'd let a migration go and (depending on the next decision), either add a warning condition for missing namespaces, or fail the migration over them (but still migrating the others). We could even add a spec flag to determine this behavior, but that may be overkill (i.e. one setting ensures src namespaces in pre-validation, and the other would just warn/fail at the end if they're not there. Then we'd need to decide which behavior would be the default.

Reply:
If we do decide to move namespace checks from pre-flight validation to during-migration checks, we'll also need to account for missing src namespaces in the "create namespaces" step. We'd want to skip creating dest namespaces if the src namespace doesn't exist.

Reply:
@dymurray Which option do you think makes more sense here? Fail early during validation phase if there's a missing src namespace, or skip namespace validation and fail the migration on a missing namespace (but still migrating the others), or just succeeding and adding a warning for missing namespaces?

Reply:
@dymurray I guess similar concerns would apply for `DirectVolumeMigration` if the src PV doesn't exist. Does that create a critical condition from validation, or do you migrate everything else and warn/fail on it?

Reply:
I think this should fail validation. Since we know its an invalid configuration we want to catch it as early as possible and prevent the user from migrating.

Reply:
Talking to @dymurray and @alaypatel07 we're leaving this as-is for now, but we may need to revisit it once the controller redesign is completed.